### PR TITLE
Autocomplete: fix `canUsePartialCompletion` for completions with leading new lines

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

In cases where multiline completions start with a leading new line, we do not account for the fact that the `truncateMultilineCompletion` call strips it. That's why `canUsePartialCompletion` returns false positives, leading to prematurely truncated multline completions.

```js
{
  "completion": {
    "text": "\nconst merge = (left, right) => {\n  let arr = [];\n  while (left.length && right.length) {",
    "newLines": 4
  },
  "truncated": {
    "text": "const merge = (left, right) => {\n    let arr = [];\n    while (left.length && right.length) {",
    "newLines": 3
  },
  "conditionForPartialReuse": "truncated.split('\n').length < completion.split('\n').length"
}
```

- [Slack thread.](https://sourcegraph.slack.com/archives/C05AGQYD528/p1697139038515929)
- A follow-up issue: https://github.com/sourcegraph/cody/issues/1387

## Test plan

Added unit tests.

### CODE REVIEW

File: vscode/src/completions/get-inline-completions-tests/streaming.test.ts [hunk 1]
```diff
@@ -134,4 +134,71 @@ describe('[getInlineCompletions] streaming', () => {
...
+          "const merge = (left, right) => {
+              let arr = [];
+              while (left.length && right.length) {
+                  if (true) {}
+              }"
+        `)
+        expect(result?.source).toBe(InlineCompletionsResultSource.Network)
+    })
+
+    it.skip('cuts-off multlineline compeltions with inconsistent indentation correctly', async () => {
```
Comment (valerybugakov) [inline vscode/src/completions/get-inline-completions-tests/streaming.test.ts:179]: This case cannot be covered by the indentation-based approach. I kept it here to re-enable this test when https://github.com/sourcegraph/cody/issues/1387 is shipped.

File: vscode/src/completions/text-processing/truncate-multiline-completion.ts [hunk 2]
```diff
@@ -36,6 +48,7 @@ export function truncateMultilineCompletion(
     if (lines.length > 1 && lines[0] === '' && indentation(lines[1]) === startIndent) {
```
Comment (philipp-spiess) [inline vscode/src/completions/text-processing/truncate-multiline-completion.ts:48]: Oh, okay, I didn't know this was the behavior (although I added it lol). I think we should split `truncateMultilineCompletion` into two distinct behaviors:

- `normalizeStartLine`
- `truncateEndBasedOnIndentation`

WDYT?

File: vscode/src/completions/can-use-partial-completion.ts [hunk 3]
```diff
@@ -23,6 +23,7 @@ export function canUsePartialCompletion(
     partialResponse: string,
     { document, multiline, docContext: { prefix, suffix } }: CanUsePartialCompletionParams
 ): boolean {
+    // return false
```
Comment (philipp-spiess) [inline vscode/src/completions/can-use-partial-completion.ts:26]:
```suggestion
```

**Review by philipp-spiess — APPROVED** [review]
Great find!

File: vscode/src/completions/text-processing/truncate-multiline-completion.ts (hunk 2 above)
Comment (valerybugakov) [reply vscode/src/completions/text-processing/truncate-multiline-completion.ts:48]: I considered it first but thought it was easier to understand the whole truncation logic in one place. How we truncated the beginning of the completion might affect the end truncation, and we might want to add new heuristics later.

Comment (philipp-spiess) [reply vscode/src/completions/text-processing/truncate-multiline-completion.ts:48]: Sounds like you gave that more thought then 👍

**Pull request merged**